code/obj/machinery/sleeper.dm

```diff
@@ -128,146 +124,96 @@
-				var/minute = round(time_left / 60)//(src.time - second) / 60
-				//DEBUG_MESSAGE("[time_of_day] - [time_left] - [minute]:[second]")
-				dat += {"<TT><B>Occupant Alarm Clock</B><br>[src.timing ? "<A href='?src=\ref[src];time=0'>Timing</A>" : "<A href='?src=\ref[src];time=1'>Not Timing</A>"] [minute]:[second]<br>
-						<A href='?src=\ref[src];tp=-30'>-</A> <A href='?src=\ref[src];tp=-1'>-</A> <A href='?src=\ref[src];tp=1'>+</A> <A href='?src=\ref[src];tp=30'>+</A><br></TT>
-						<br><font size=2>System will inject rejuvenators automatically when occupant is in hibernation.</font><hr
-						<A href='?src=\ref[src];refresh=1'>Refresh</A> | <A href='?src=\ref[src];rejuv=1'>Inject Rejuvenators</A> | <A href='?src=\ref[src];eject_occupant=1'>Eject Occupant</A>"}
-
-			else
-				dat += "<HR>The sleeper is unoccupied."
-
-		else
-			dat += {"<font color='red'><b>ERROR:</b> No sleeper detected!</font><br>
-					<br><A href='?src=\ref[src];refresh=1'>Refresh Connection</A>"}
-
-		user.Browse(dat, "window=sleeper")
-		onclose(user, "sleeper")
-
-		return
-
-	Topic(href, href_list)
-		/*if (..())
-			return*/
-		if (!isturf(src.loc) && !istype(src.loc, /obj/machinery/sleeper))
-			return
-		if ((src.our_sleeper && src.our_sleeper.occupant == usr) || usr.getStatusDuration("stunned") || usr.getStatusDuration("weakened") || usr.stat || usr.restrained())
-			return
-		if (!issilicon(usr) && !in_range(src, usr))
+	ui_status(mob/user)
+		var/use_obj = src
+		if (src.loc == src.our_sleeper) // port-a-medbay
+			use_obj = src.loc
+			if (user in use_obj)
+				return UI_CLOSE
+		return min(
+			tgui_broken_state.can_use_topic(use_obj, user),
+			tgui_physical_state.can_use_topic(use_obj, user),
+			tgui_not_incapacitated_state.can_use_topic(use_obj, user)
+		)
+
+	ui_act(action, list/params, datum/tgui/ui, datum/ui_state/state)
+		. = ..()
+		if (.)
 			return
-
-		src.add_fingerprint(usr)
-		src.add_dialog(usr)
-
-		if (href_list["time"])
-			if (src.our_sleeper && src.our_sleeper.occupant)
-				if (isdead(src.our_sleeper.occupant))
-					usr.show_text("The occupant is dead.", "red")
-				else
-					src.timing = text2num(href_list["time"])
+		switch(action)
+			if("timer")
+				if (src.our_sleeper && src.our_sleeper.occupant && !isdead(src.our_sleeper.occupant))
+					src.timing = !src.timing
 					src.visible_message("<span class='notice'>[usr] [src.timing ? "sets" : "stops"] the [src]'s occupant alarm clock.</span>")
 					if (src.timing)
-						src.time_started = world.timeofday//realtime
+						src.time_started = TIME
 						// People do use sleepers for grief from time to time.
 						logTheThing("station", usr, src.our_sleeper.occupant, "initiates a sleeper's timer ([src.our_sleeper.emagged ? "<b>EMAGGED</b>, " : ""][src.time/10] seconds), forcing [constructTarget(src.our_sleeper.occupant,"station")] asleep at [log_loc(src.our_sleeper)].")
 					else
+						src.time = clamp(src.time + src.time_started - TIME, 0, 1800)
```

Comment:
```suggestion
						src.time = clamp(src.time + src.time_started - TIME, 0, 3 MINUTES)
```

Reply:
Actually, as we're re-using this all over the place, maybe make it a `var` (`var/maximum_time = 3 MINUTES` or whatever)?